src/main/scala/io/iohk/praos/domain/VirtualGenesis.scala

```diff
@@ -0,0 +1,37 @@
+package io.iohk.praos.domain
+import io.iohk.praos.util.Logger
+
+trait VirtualGenesis {
+
+  def computeGenesisForEpoch(slotInEpoch: SlotInEpoch, genesisHistory: GenesisHistory): Option[Genesis]
+}
+
+case class VirtualGenesisImpl(epochLength: Int, lengthForCommonPrefix: Int) extends VirtualGenesis with Logger {
+
+  override def computeGenesisForEpoch(slotInEpoch: SlotInEpoch, genesisHistory: GenesisHistory): Option[Genesis] = {
+    val computeGenesis = (computeGenesisForStaticCase orElse computeGenesisForDynamicCase).lift
+    computeGenesis(slotInEpoch, genesisHistory)
+  }
+
+  /**
+    * The paper refer to the first epoch as the static case of the protocol.
+    * @return the initial genesis
+    */
+  private val computeGenesisForStaticCase: PartialFunction[(SlotInEpoch, GenesisHistory), Genesis] = {
+    case (slotInEpoch, genesisHistory) if slotInEpoch.epochNumber == 1 => genesisHistory.getGenesisAt(0).get
```

Comment:
Maybe it would be better to remove both this `.get` and the one from `computeGenesisForDynamicCase` by having this partial funcitons return an `Option[Genesis]`?

Reply:
I'm agree too =)